Python/chapter02/2.5 - Sum Lists/miguel_2.5_soln.py

```diff
@@ -0,0 +1,447 @@
+        return a is None and b is None
+
+
+def sum_lists_forward(ll1: LinkedList, ll2: LinkedList) -> LinkedList:
+    """
+    Will add ll1 and ll2 where each element in the linked lists
+    represent digits.
+    Note: Each element in the input linked lists MUST be values
+    between 0-9
+    Digits are in forward order, for example:
+    Input: (6 -> 1 -> 7) + (2 -> 9 -> 5). That is, 617 + 295
+    Output: 9 -> 1 -> 2. That is, 912
+    Runtime: O(n)
+    Space Complexity: O(n) - output_ll
+    :param ll1: first input linked list
+    :param ll2: second input linked list
+    :return: a linked list containing the result of the addition
+    """
+    shorter_ll, longer_ll = (ll1, ll2) if ll1.size < ll2.size else (ll2, ll1)
+    n_shorter = shorter_ll.head
+    n_longer = longer_ll.head
+    length_diff = abs(shorter_ll.size - longer_ll.size)
+    # first, advance longer list to line up with
+    # digit of shorter list
+    # for example, if we have:
+    # longer_ll = 9 -> 9 -> 9 -> 9
+    # shorter_ll = 9 -> 9
+    # we will line up the current pointer
+    # of longer_ll to correspond to the
+    # '10s' digit place because we must add
+    # digits that are in the same place value.
+    # n_longer will point to the following parenthesis
+    # 9 -> 9 -> (9) -> 9
+    # n_shorter will point to the beginning of shorter_ll
+    # (9) -> 9
+    for i in range(length_diff):
+        n_longer = n_longer.next
+    output_ll = LinkedList()
+    # perform addition, taking into account a
+    # 'backwards' carry value since we
+    # are looping forward digits down to
+    # 'ones' digit place
+    prev_node = None
+    leading_bw_carry = False
+    while n_shorter is not None:
+        first = n_shorter.data
+        second = n_longer.data
+        result = first + second
+        if result >= 10:
+            backwards_carry = 1
+            result -= 10  # extract digit in one's place
+            if prev_node is None:
+                # for when the first addition yields a carry
+                # and the output linked list is empty
+                leading_bw_carry = True
+            else:
+                # add carry to previous digit
+                prev_node.data += backwards_carry
+        output_ll.append_to_tail(result)
+        n_shorter = n_shorter.next
+        n_longer = n_longer.next
+        prev_node = output_ll.tail
+
+    # first case after adding same digit places,
+    # both input lists same size
+    if length_diff == 0:
+        if leading_bw_carry:
+            backwards_carry = 1
+            output_ll.append_to_head(backwards_carry)
+        return output_ll
+    # otherwise, we have non-equal lengths
+    # with or without a leading backward carry
+    # leading bw carry means the the size diff
+    # decreased.
+    # no leading bw carry means size diff stayed the same
+    if leading_bw_carry:
+        backwards_carry = 1
+    else:
+        backwards_carry = 0
+    # otherwise, as long as the diff is >= 0
```

Comment:
This code is kinda complicated and not too easy to follow since it has a lot of edge cases :( In order to coalesce all the edge cases together, would it be simpler if you allowed yourself to mutate the output linked list before returning it? Either the values and/or the order of the nodes within the linked list.

That way you can deal with the core functionality at first (finding the first shared node, then adding both lists), and you can push the carry logic to a second pass.

Reply:
> In order to coalesce all the edge cases together, would it be simpler if you allowed yourself to mutate the output linked list before returning it? Either the values and/or the order of the nodes within the linked list.

What do you mean by mutating the output linked list?  Do you mean after the core functionality, then I can pad the output list with 0s?  That approach would be cleaner.  The second pass would handle that.

>
It does add an extra pass, but the overall time / space complexity would be unchanged. In any case, most problems that deal with linked lists are sort of artificial: in a realistic scenario one would change the linked list to a better data structure instead of doing all these contortions, so I wouldn't feel too bad about the execution time / code complexity trade-off.

Yeah, if this was a doubly linked list, it would have been a bit more straightforward. (or rather, straightbackward ;))

Reply:
> > In order to coalesce all the edge cases together, would it be simpler if you allowed yourself to mutate the output linked list before returning it? Either the values and/or the order of the nodes within the linked list.
> 
> What do you mean by mutating the output linked list? Do you mean after the core functionality, then I can pad the output list with 0s? That approach would be cleaner. The second pass would handle that.

Not sure how padding the output list with zeroes would help, but if it leads to simpler code, go for it. There's also the option of adding stuff to the nodes without taking care of the carry and deferring the handling of carry to afterwards.

Whatever modifications you do to the output list are implementation details and are fair game, as long as the return value is correct.

> 
> > It does add an extra pass, but the overall time / space complexity would be unchanged. In any case, most problems that deal with linked lists are sort of artificial: in a realistic scenario one would change the linked list to a better data structure instead of doing all these contortions, so I wouldn't feel too bad about the execution time / code complexity trade-off.
> 
> Yeah, if this was a doubly linked list, it would have been a bit more straightforward. (or rather, straightbackward ;))

you can invert a singly-linked list in O(n) time O(1) space if you need to.

Reply:
Would it be fair game to make reversed copies of the input linked lists and then call the sum_lists function on them?

Reply:
totally!

Reply:
Wow lol I should have done that from the beginning but I thought it was too simple

Reply:
simplicity is ranked pretty high in the list of desirable properties of software. but simplicity is not always free, since it sometimes comes with a trade-off of time or space. you often need to at least enumerate all your options and choose whatever is best in that instance.